index.js

```diff
@@ -14,47 +13,90 @@ function createResolver(config, options) {
     resolve: resolve,
   };
 
-  var lastKey;
+  // Keep constants separately
+  var constants = {};
+
+  // Keep requested keys to detect (and disallow) recursive resolution
+  var stack = [];
 
   function resolve(key) {
-    if (typeof key !== 'string') {
-      return;
-    }
 
-    var appliedArgs = slice.call(arguments, 1);
+    if (constants.hasOwnProperty(key)) {
+      return constants[key];
+    }
 
     var definition = config[key];
     // Ignore options that are not defined
     if (!definition) {
       return;
     }
 
-    if (key === lastKey) {
+    if (stack.some(function(s) {
+      return s === key;
+    })) {
       throw new Error('Recursive resolution denied.');
     }
-    lastKey = key;
 
-    var option = options[key];
-    // Bind the option so it can resolve other options if necessary
-    if (typeof option === 'function') {
-      option = option.bind(resolver);
+    stack.push(key);
+    try {
+      var option = options[key];
+      var appliedArgs = slice.call(arguments, 1);
+      var args = [definition.type, option].concat(appliedArgs);
+
+      if (typeof option === 'function') {
+        option = normalize.apply(resolver, args);
+      } else {
+        option = undefined;
+      }
+
+      if (option === undefined) {
```

Comment:
Ah yes, yes it should. Leftover from my suggestion earlier about having value-or-function use strictly undefined for rejection.